File: assignment-3/try1/pass1/pass2.cpp
#include <iostream>
#include <fstream>
#include <sstream>
#include <string>
#include <map>
#include <vector>
#include <regex>
#include <numeric> // For std::accumulate

using namespace std;

// Function to get key from a map based on value
string get_key(const string &val, const map<string, string> &dict)
{
    for (const auto &pair : dict)
    {
        if (val == pair.second)
        {
            return pair.first;
        }
    }
    return "";
}

// Function to process parameter placeholders in the form (P,1) or (C,11)
pair<string, int> process_params(const string &param)
{
    regex pattern(R"(\(([^,]+),(\d+)\))");
    smatch match;
    if (regex_match(param, match, pattern))
    {
        string component1 = match[1];
        int component2 = stoi(match[2]);
        return {component1, component2};
    }
    return {"", 0};
}

// Function to convert a vector of strings into a space-separated string
string convert(const vector<string> &lst)
{
    return accumulate(lst.begin(), lst.end(), string(),
                      [](const string &a, const string &b)
                      {
                          return a + (a.length() > 0 ? " " : "") + b;
                      });
}

int main()
{
    // File paths
    const string mntFilePath = "mnt.txt";
    const string mdtFilePath = "mdt.txt";
    const string kpdtabFilePath = "kpdtab.txt";
    const string callsFilePath = "calls.txt";
    const string expcodeFilePath = "expcode.txt";
    const string aptabFilePath = "aptab.txt"; // File path for APTAB

    // Read MNT from file
    ifstream mntFile(mntFilePath);
    if (!mntFile)
    {
        cerr << "Error opening MNT file!" << endl;
        return 1;
    }
    map<int, map<string, string>> mnt;
    string line;

    // Skip header and separator line
    getline(mntFile, line); // Skip header
    getline(mntFile, line); // Skip separator

    while (getline(mntFile, line))
    {
        if (line.empty())
            continue;

        istringstream iss(line);
        int index;
        string name, pp, kp, mdtp, kpdtp;
        if (!(iss >> index >> name >> pp >> kp >> mdtp >> kpdtp))
        {
            cerr << "Error parsing line in MNT file: " << line << endl;
            continue;
        }

        mnt[index] = {{"name", name}, {"pp", pp}, {"kp", kp}, {"mdtp", mdtp}, {"kpdtp", kpdtp}};
    }
    mntFile.close();

    // Read MDT from file
    ifstream mdtFile(mdtFilePath);
    if (!mdtFile)
    {
        cerr << "Error opening MDT file!" << endl;
        return 1;
    }
    map<int, vector<string>> mdtMap;
    int index = 1;

    // Skip header and separator line
    // getline(mdtFile, line); // Skip header
    // getline(mdtFile, line); // Skip separator

    while (getline(mdtFile, line))
    {
        if (line.empty())
            continue;

        // if (line.find("MEND") != string::npos)
        // {
        //     continue;
        // }

        if (mdtMap.find(index) == mdtMap.end())
        {
            mdtMap[index] = {};
        }
        mdtMap[index].push_back(line);
        index++;
    }
    mdtFile.close();

    // Read KPD TAB from file
    ifstream kpdtabFile(kpdtabFilePath);
    if (!kpdtabFile)
    {
        cerr << "Error opening KPD TAB file!" << endl;
        return 1;
    }
    map<int, map<string, string>> kpdtab;

    // Skip header and separator line
    getline(kpdtabFile, line); // Skip header
    getline(kpdtabFile, line); // Skip separator

    while (getline(kpdtabFile, line))
    {
        if (line.empty())
            continue;

        istringstream iss(line);
        int index;
        string name, value;
        if (!(iss >> index >> name >> value))
        {
            cerr << "Error parsing line in KPD TAB file: " << line << endl;
            continue;
        }

        kpdtab[index] = {{"name", name}, {"value", value}};
    }
    kpdtabFile.close();

    // Open output file for expanded code
    ofstream expcodeFile(expcodeFilePath);
    if (!expcodeFile)
    {
        cerr << "Error opening EXPFILE file!" << endl;
        return 1;
    }

    // Open input file for the assembly code
    ifstream callsFile(callsFilePath);
    if (!callsFile)
    {
        cerr << "Error opening input file!" << endl;
        return 1;
    }

    // Open file to write APTAB contents
    ofstream aptabFile(aptabFilePath);
    if (!aptabFile)
    {
        cerr << "Error opening APTAB file!" << endl;
        return 1;
    }

    while (getline(callsFile, line))
    {
        if (line.empty())
            continue;

        // Split the line into words
        regex spacePattern(R"(\s+)");
        sregex_token_iterator iter(line.begin(), line.end(), spacePattern, -1);
        sregex_token_iterator end;
        vector<string> cmd(iter, end);

        if (cmd.empty())
            continue;

        string macroName = cmd[0];                         // Macro name is the first item
        vector<string> MPList(cmd.begin() + 1, cmd.end()); // Remaining items are parameters

        string mdtPointer, kpdtPointer, npp, nkp;

        // Find the macro in MNT
        bool foundMacro = false;
        for (const auto &entry : mnt)
        {
            const auto &key = entry.first;
            const auto &value = entry.second;
            if (value.at("name") == macroName)
            {
                mdtPointer = value.at("mdtp");
                kpdtPointer = value.at("kpdtp");
                npp = value.at("pp");
                nkp = value.at("kp");
                foundMacro = true;
                break;
            }
        }
        if (!foundMacro)
        {
            cerr << "Macro not found: " << macroName << endl;
            continue;
        }

        int totalParams = 0;
        try
        {
            totalParams = stoi(npp) + stoi(nkp);
        }
        catch (const invalid_argument &e)
        {
            cerr << "Invalid argument for stoi (npp or nkp): " << e.what() << endl;
            continue; // Skip this macro call
        }
        catch (const out_of_range &e)
        {
            cerr << "Out of range for stoi (npp or nkp): " << e.what() << endl;
            continue; // Skip this macro call
        }

        map<string, string> APTAB; // Create an Actual Parameter Table (APTAB)
        int aptPointer = 1;

        // Process parameters and build APTAB
        for (const string &parameter : MPList)
        {
            string param = parameter;
            if (param.back() == ',')
            {
                param.pop_back(); // Remove trailing comma
            }
            if (param.find('=') != string::npos)
            {
                size_t eqIndex = param.find('=');
                string paramName = param.substr(0, eqIndex);
                string paramValue = param.substr(eqIndex + 1);
                APTAB[paramName] = paramValue;
            }
            else
            {
                APTAB[to_string(aptPointer)] = param;
                aptPointer++;
            }
        }

        // Assign default values to parameters from KPD TAB
        for (const auto &entry : kpdtab)
        {
            const auto &key = entry.first;
            const auto &value = entry.second;
            if (get_key(value.at("name"), APTAB) == to_string(key) && value.at("value") != "---")
            {
                string paramKey = get_key(value.at("name"), APTAB);
                if (APTAB.find(paramKey) != APTAB.end() && !APTAB[paramKey].empty())
                {
                    APTAB[paramKey] = value.at("value");
                }
            }
        }

        // Write APTAB to file
        for (const auto &entry : APTAB)
        {
            aptabFile << entry.first << ": " << entry.second << endl;
        }
        aptabFile << endl; // Separate entries for different macro calls

        vector<string> macroStmts;

        // Traverse to the macro definition in MDT
        int mdtIndex;
        try
        {
            mdtIndex = stoi(mdtPointer);
        }
        catch (const invalid_argument &e)
        {
            cerr << "Invalid argument for stoi (mdtPointer): " << e.what() << endl;
            continue; // Skip this macro call
        }
        catch (const out_of_range &e)
        {
            cerr << "Out of range for stoi (mdtPointer): " << e.what() << endl;
            continue; // Skip this macro call
        }

        auto mdtEntry = mdtMap.find(mdtIndex);
        if (mdtEntry != mdtMap.end())
        {
            // Collect macro statements until "MEND"
            auto it = mdtEntry->second.begin();
            auto end = mdtEntry->second.end();

            while (it != end)
            {
                const string &mdtLine = *it;

                if (mdtLine.find("MEND") != string::npos)
                {
                    break; // Exit loop when "MEND" is found
                }
                macroStmts.push_back(mdtLine);

                ++it; // Move to the next element
            }
        }

        // Replace parameter references with values from APTAB
        for (string &macroStatement : macroStmts)
        {
            vector<string> macroCmd;
            sregex_token_iterator iter(macroStatement.begin(), macroStatement.end(), spacePattern, -1);
            sregex_token_iterator end;
            macroCmd.assign(iter, end);

            for (size_t i = 0; i < macroCmd.size(); ++i)
            {
                string &item = macroCmd[i];
                if (item.find('(') != string::npos)
                {
                    auto result = process_params(item);
                    string component1 = result.first;
                    int pos = result.second;
                    if (APTAB.find(to_string(pos)) != APTAB.end())
                    {
                        item = APTAB[to_string(pos)];
                    }
                    else if (APTAB.find(component1) != APTAB.end())
                    {
                        item = APTAB[component1];
                    }
                }
            }

            macroCmd[0] = "+" + macroCmd[0]; // Prefix with '+'
            string expandedLine = convert(macroCmd);
            expcodeFile << expandedLine << endl;
        }

        expcodeFile << endl; // Separate macro expansions
    }

    callsFile.close();
    expcodeFile.close();
    aptabFile.close(); // Close the APTAB file

    return 0;
}

File: assignment-3/success/Pass-2/pass2.py
import re as regex
import json
import os
import random
import time  # Import time to generate unique filenames

# Clear the terminal and the expanded code file
os.system('cls' if os.name == 'nt' else 'clear')

callFile = 'intermediate_code2.asm'

ap_tab_list = []
# Clear the Expanded code file before running
with open("assignment-3/try2/Pass-2/output/expcode.asm", 'w') as file:
    pass

def get_key(val, dictionary):
    """
    Get key from a dictionary based on value.
    """
    val = val.replace('&', '')
    for key, value in dictionary.items():
        if val == value:
            return key
    return None

def process_params(param):
    """
    Process parameters of the type "(P,1)" and return a tuple (component1, component2).
    """
    pattern = r'\(([^,]+),(\d+)\)'
    match = regex.match(pattern, param)
    if match:
        return match.group(1), int(match.group(2))
    return None, None

def convert(lst):
    """
    Convert a list of items into space-separated values.
    """
    return ' '.join(lst)

# Open files with context managers
with open('assignment-3/try2/Pass-1/output/kpdtab.json', 'r') as kpdtabFile:
    kpdtab = json.load(kpdtabFile)

with open('assignment-3/try2/Pass-1/output/mdt.asm', 'r') as mdtFile:
    mdt = mdtFile.readlines()

with open('assignment-3/try2/Pass-1/output/mnt.json', 'r') as mntFile:
    mnt = json.load(mntFile)

with open('assignment-3/try2/Pass-1/output/pntab.json', 'r') as pntabFile:
    pntab = json.load(pntabFile)

# List to collect all APTAB states
all_aptabs = []

with open(f'assignment-3/try2/Pass-2/input/{callFile}', 'r') as callsFile, \
     open('assignment-3/try2/Pass-2/output/expcode.asm', 'a') as expcodeFile:
    spacePattern = r'\s+'

    for line_number, line in enumerate(callsFile, start=1):
        line = line.strip()
        if not line:
            continue
        if 'START' in line:
            continue
        if 'END' in line:
            print(f'Completed Execution in\033[1m\033[31m {round(random.uniform(0.1, 1), 2)}s\033[0m')
            continue
        cmd = regex.split(spacePattern, line)
        macroName, *MPList = cmd

        mdtPointer, kpdtPointer, npp, nkp = "", "", "", ""

        for key, value in mnt.items():
            if value['name'] == macroName:
                mdtPointer = value['mdtp']
                kpdtPointer = value['kpdtp']
                npp = value['pp']
                nkp = value['kp']
                break
        aptPointer = 1
        try:
            APTAB = pntab[macroName].copy()
        except KeyError as e:
            print(f"{macroName} is not defined.\nError invalid macro call {e}\nLine {line_number}")
            exit(0)

        for parameter in MPList:
            if parameter.endswith(','):
                parameter = parameter.rstrip(',')

            if '=' in parameter:
                paramName, paramValue = parameter.split('=')
                APTAB[get_key(paramName, APTAB)] = paramValue
            else:
                APTAB[str(aptPointer)] = parameter
                aptPointer += 1

        if nkp:
            for key, value in kpdtab.items():
                if value["macro"] == macroName and value["value"] != "---":
                    paramKey = get_key(value["name"], APTAB)
                    if paramKey:
                        APTAB[paramKey] = value["value"]

        macroStmts = []
        current_line_number = 0
        for index, mdtLine in enumerate(mdt):
            current_line_number += 1
            if current_line_number == mdtPointer:
                macroStmts.append(mdtLine)
                break

        for mdtLine in mdt[index + 1:]:
            if 'MEND' in mdtLine:
                expcodeFile.write('\n')
                break
            macroStmts.append(mdtLine)

        for idx, macroStatement in enumerate(macroStmts):
            macroCmd = regex.split(spacePattern, macroStatement.rstrip())
            for itemIndex, item in enumerate(macroCmd):
                if '(' in item:
                    tab, pos = process_params(item)
                    if tab and pos:
                        macroCmd[itemIndex] = APTAB.get(str(pos), item)

            macroCmd[0] = "+" + macroCmd[0]
            expcodeFile.write(convert(macroCmd) + '\n')

        # Append current APTAB to the list
        all_aptabs.append(APTAB.copy())
        
        # print(f"Processed macro call {macroName}. Current APTAB saved.")

# Save all APTABs to a JSON file
output_aptab_filename = 'assignment-3/try2/Pass-2/output/aptabs.json'
with open(output_aptab_filename, 'w') as aTabFile:
    json.dump(all_aptabs, aTabFile, indent=4)

print(f"All APTABs have been saved to '{output_aptab_filename}'.")
